File: 811. Subdomain Visit Count/_811.py
class Solution:
    def subdomainVisits(self, cpdomains):
        myDict = {}
        for i in cpdomains:
            v, domain = i.split()
            v = int(v)
            tmp = ""
            for j in domain.split('.')[::-1]:
                tmp = j + tmp
                tmpv = myDict.get(tmp, 0)
                myDict[tmp] = tmpv+v
                tmp = "." + tmp
        return [str(myDict[i])+ " "+ i for i in myDict]

    def subdomainVisits1(self, cpdomains):
        import collections
        counter = collections.Counter()
        for cpdomain in cpdomains:
            count, *domains = cpdomain.replace(" ",".").split(".")
            for i in range(len(domains)):
                counter[".".join(domains[i:])] += int(count)
        return [" ".join((str(v), k)) for k, v in counter.items()]

File: 811. Subdomain Visit Count/811.go
package leetcode

import (
	"strings"
	"strconv"
)

func subdomainVisits(cpdomains []string) []string {
	myDict := make(map[string]int, 1)
	for _, v := range cpdomains{
		tmp := strings.Split(v, " ")
		value, _ := strconv.Atoi(tmp[0])
		domain := strings.Split(tmp[1], ".")

		tmpstr := ""
		for i:=len(domain)-1; i>-1; i--{
			v2 := domain[i]
			tmpstr = v2 + tmpstr
			myDict[tmpstr] += value
			tmpstr = "." + tmpstr
		}
	}
	
	res := make([]string, 0 ,len(myDict))
	for i, v := range myDict{
		res = append(res, strconv.Itoa(v)+" "+i)
	}
	return res
}

// 这个自己实现split要快很多 测试显示动态开辟内存次数会变少？？？
func subdomainVisits1(cpdomains []string) []string {
	tempSlice := []string{}
	tempMap := make(map[string]int)
	var n int
	var s string
	for _, v := range cpdomains {
		for I, V := range v {
			if V == ' ' {
				n, _ = strconv.Atoi(v[:I])
				s = v[I+1:]
			}
		}
		tempMap[s] += n
		for i, v := range s {
			var tempS string
			if v == '.' {
				tempS = s[i+1:]
				tempMap[tempS] += n
			}
		}
	}
	for i, v := range tempMap {
		a := strconv.Itoa(v) + " " + i
		tempSlice = append(tempSlice, a)
	}
	return tempSlice
}